### PR TITLE
Fix: Github api cache issues

### PR DESCRIPTION
**Pull request by ZibanPirate**
# Description

These changes will fix the issue with `./api` consuming more Github API quota then it needs, due to two reasons:
- [this project](https://github.com/dzcode-io/dzcode.io/pull/426/files#diff-01aa53b874cfcbf7c35edde44271bf897a4eec2018c42638044391c7f4902b6b) got renamed, and Github now returns 301 which consume API Rate Limit.
- Calls to Github API happen concurrently, which may lead to potential corruption of cache data (eg: two calls happen at the same time, and both try to update the cache)

The changes include:
- Added a non-blocking [`lock`](https://github.com/dzcode-io/dzcode.io/pull/426/files#diff-c87f7bb1e49f256f08c6865cbbbb55660edd0f8667f0c91023f5c9d896287b71R8) function:
  - you give it another function as an argument, and It returns a [locked](https://en.wikipedia.org/wiki/Lock_(computer_science)) version of that function.
  - the locked function ensures that it only runs once at a time.
  - If you call the locked function multiple times while it's still in use (eg: in a `Promise.all`), these new calls will be added to a [queue](https://github.com/dzcode-io/dzcode.io/pull/426/files#diff-c87f7bb1e49f256f08c6865cbbbb55660edd0f8667f0c91023f5c9d896287b71R9), and will get executed once their turns come.
  - If you call the locked function multiple times **with the same arguments** while it's still in use, here it will not add the new calls to the queue, instead, it will attach them to the existent queue Item, and then when this Item's turn is reached, it will resolve and call back all the attached function calls, both the old ones and the newly added calls.
- Made the `FetchService` more versatile, allowing to pass headers.
- Added optional `GITHUB_TOKEN` params, which will be used to increase the Rate Limit from 60 to 5000.
- Fixed the redirection response we get because of the renaming of the `Moadaly` project.
- Added a Github controller with `/Github/RateLimit` endpoint
- Added a script to generate nodemon.json to watch all `@dzcode.io/*` dependencies, and used it in `./api` and `./data`

## Type of change

- Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: data/package.json [hunk 1]
```diff
@@ -52,7 +52,7 @@
     "lint:eslint": "eslint --config ../packages/tooling/.eslintrc.json --ignore-path ../packages/tooling/.eslintignore --report-unused-disable-directives",
     "lint:prettier": "prettier  --config ../packages/tooling/.prettierrc --ignore-path ../packages/tooling/.prettierignore --loglevel warn",
     "start": "node dist/index.js",
-    "start:dev": "nodemon dist/index.js",
+    "start:dev": "ts-node ../packages/tooling/nodemon.ts @dzcode.io/api && nodemon dist/index.js",
```
Comment (ZibanPirate) [inline data/package.json:55]:
```suggestion
    "start:dev": "ts-node ../packages/tooling/nodemon.ts @dzcode.io/data && nodemon dist/index.js",
```

File: README.md [hunk 2]
```diff
@@ -68,6 +68,14 @@ yarn dev:all
 - For api server go to <http://localhost:7070/docs>
 - For mobile go to <http://localhost:19002> and scan QR code with your phone
 
+**Note**
+
+In `./api`, keep in mind that you have limited calls to Github Api (60 calls per hour), the FetchService is doing a great job at caching theses calls so it doesn't unnecessarily consume Github API quota. If you wish to extend the limit from 60 to 5000, simply create a [Github Personal Access Token](https://github.com/settings/tokens) (make sure it has `Access public repositories` checked), and set it in `./api/.env` like this:
```
Comment (ZibanPirate) [inline README.md:73]:
```suggestion
In [`./api`](./api), keep in mind that you have limited calls to Github Api (60 calls per hour), the [FetchService](./api/src/fetch/service.ts) is doing a great job at caching theses calls so it doesn't unnecessarily consume Github API quota. If you wish to extend the limit from 60 to 5000, simply create a [Github Personal Access Token](https://github.com/settings/tokens) (make sure it has `Access public repositories` checked), and set it in `./api/.env` like this:
```

Comment (ZibanPirate) [reply README.md]: 81a5621

File: data/package.json (hunk 1 above)
Comment (ZibanPirate) [reply data/package.json]: 81a5621

**Review by Omar-Belghaouti — APPROVED** [review]

**Pull request merged**